lib/spack/spack/url_buildcache.py

```diff
@@ -0,0 +1,1129 @@
+        return BuildcacheManifest(
+            layout_version=manifest_json["version"],
+            data=[BlobRecord.from_json(blob_json) for blob_json in manifest_json["data"]],
+        )
+
+    def get_blob_records(self, content_type: str) -> List[BlobRecord]:
+        matches: List[BlobRecord] = []
+
+        for record in self.data:
+            if record.content_type == content_type:
+                # matches.append(BlobRecord.from_json(record.to_json()))
+                matches.append(record)
+
+        if matches:
+            return matches
+
+        raise NoSuchBlobException(f"Manifest has no blobs of type {content_type}")
+
+
+class URLBuildcacheEntry:
+    """A class for managing URL-style buildcache entries
+
+    This class manages access to a versioned buildcache entry by providing
+    a means to download both the metadata (spec file) and compressed archive.
+    It also provides methods for accessing the paths/urls associcated with
+    buildcache entries.
+
+    Starting with buildcache layout version 3, it is not possible to know
+    the full path to a compressed archive without either building it locally,
+    or else fetching and reading the metadata first.  This class provides api
+    for fetching the metadata, as well as fetching the archive, and it enforces
+    the need to fetch the metadata first.
+
+    To help with downloading, this class manages two spack.spec.Stage objects
+    internally, which must be destroyed when finished.  Specifically, if you
+    call either of the following methods on an instance, you must eventually also
+    call destroy():
+
+        fetch_metadata()
+        fetch_archive()
+
+    """
+
+    SPEC_URL_REGEX = re.compile(r"(.+)/v([\d]+)/manifests/.+")
+    LAYOUT_VERSION = 3
+    INDEX_VERSION = f"index-v{spack.database._DB_VERSION}"
+    SPEC_VERSION = f"spec-v{spack.spec.SPECFILE_FORMAT_VERSION}"
+    TARBALL_VERSION = "tarball-v1"
+    PUBLIC_KEY = "key-v1"
+    PUBLIC_KEY_INDEX = "key-index-v1"
+    COMPONENT_PATHS = {
+        BuildcacheComponent.BLOBS: ["blobs"],
+        BuildcacheComponent.INDICES: [f"v{LAYOUT_VERSION}", "manifests"],
+        BuildcacheComponent.INDEX: [f"v{LAYOUT_VERSION}", "manifests", INDEX_MANIFEST_FILE],
+        BuildcacheComponent.KEYS: [f"v{LAYOUT_VERSION}", "manifests"],
+        BuildcacheComponent.SPECS: [f"v{LAYOUT_VERSION}", "manifests"],
+        BuildcacheComponent.LAYOUT_JSON: [f"v{LAYOUT_VERSION}", "layout.json"],
+    }
+
+    def __init__(self, mirror_url: str, spec: Optional[spack.spec.Spec] = None):
+        """Lazily initialize the object"""
+        self.mirror_url: str = mirror_url
+        self.spec: Optional[spack.spec.Spec] = spec
+        self.manifest: Optional[BuildcacheManifest] = None
+        self.remote_manifest_url: str = ""
+        self.stages: Dict[BlobRecord, spack.stage.Stage] = {}
+
+    @classmethod
+    def get_layout_version(cls) -> int:
+        return cls.LAYOUT_VERSION
+
+    @classmethod
```

Comment:
bit confused why there are dozens of classmethods, can't they be free functions?

Reply:
Some of the class methods access class variables that are defined differently in the derived class so we get some layout-appropriate polymorphism.  Some of the classmethods are overridden in the derived class for one reason or another (in the case of the v2 class, it's usually to raise, indicating that operation isn't supported for v2). I have also made classmethods whenever it seemed to me that the behavior could be specific to a layout version.